### ISSUE TITLE
feat: would like patch numbers to match between iOS and Android

### ISSUE THREAD

**Issue by eseidel** [shorebirdtech/shorebird#1382]
Right now if you publish a patch to Android and then a patch to iOS (for the same app), they will end up with patch number "1" and "2" respectively.  This is confusing to some users who expect the patch numbers to match for the "same patch".  This also causes some complexity for us in the console and could make sense to want to allow users to just mark one patch live and have it affect all OSes.

**Comment by felangel** [comment]
Agreed this should be fairly straightforward to implement since it’s only a CLI change.

**Comment by anderscheow** [comment]
Just for my use case:

App is showing version + patch in About screen. For example 1.0.1 (13) - Patch 1

Current behaviour:
Android: 1.0.1 (13) - Patch 1
iOS: 1.0.1 (13) - Patch 2

Expected behaviour:
Android: 1.0.1 (13) - Patch 1
iOS: 1.0.1 (13) - Patch 1

**Comment by anderscheow** [comment]
@eseidel @felangel any idea when this will get implemented?

**Comment by eseidel** [comment]
We currently have no schedule for this, no.  We were discussing something similar yesterday which is that these numbers are currently just "names" and the fact that we have them auto-increment on a per-platform basis is confusing in other contexts.  We could easily change the name to something else.  The underlying "id" of the patch is just our database id for the patch artifact table row.

**Comment by eseidel** [comment]
Dupe with https://github.com/shorebirdtech/shorebird/issues/1762?

**Comment by alejojarahi** [comment]
Hi, I did not want to generate different patches per platform, i was testing the functionality of using `shorebird patch --platforms android,ios`  https://github.com/shorebirdtech/shorebird/issues/1762 but it doesn't work well if to use iOS exclusive flags, for example --export-options-plist so I don't know if this is a bug, something that is not covered, or am I doing something wrong, thanks.

@eseidel @felangel 

```
shorebird patch --platforms=android,ios --flavor=dev --release-version=1.0.0 --track=beta --export-options-plist=ios/ExportOptions.plist --no-confirm
```

Shorebird version: 1.5.5
Console error:

```
⠋ Fetching apps...
✓ Fetching apps (0.4s)
⠋ Fetching releases...
✓ Fetching releases (0.2s)
⠋ Fetching aab artifact...
✓ Fetching aab artifact (0.2s)
⠋ Downloading aab...
⠙ Downloading aab (0%)... (0.2s)
⠹ Downloading aab (9%)... (0.4s)
⠸ Downloading aab (24%)... (0.7s)
⠼ Downloading aab (43%)... (0.9s)
⠴ Downloading aab (63%)... (1.2s)
⠦ Downloading aab (82%)... (1.4s)
✓ Downloading aab (100%) (1.7s)
⠋ Building patch with Flutter 3.24.5 (3faf56aafa)...
✗ Failed to build: [  +13 ms] Could not find an option named "export-options-plist".

[   +1 ms] Run 'flutter -h' (or 'flutter <command> -h') for available flutter commands and options. (6.2s)

If you aren't sure why this command failed, re-run with the --verbose flag to see more information.

```

**Comment by bryanoltman** [comment]
@alejojarahi This doesn't seem related to this bug. Would you mind opening a new one and clearly describing your problem there?

**Comment by alexobviously** [comment]
Any chance of this being fixed soon? This bug causes some confusion and makes communicating with customers a bit harder than it ideally should be. "The changes are live, make sure you're on patch 3 on your android devices, and patch 4 on ios" <- not the best.

Would be much appreciated. If this is not done soon, I need to build some sort of workaround like displaying `patch - 1` for ios, but I would prefer not to do that.

**Comment by eseidel** [comment]
I appreciate the feedback.  It's not on our near-term roadmap at the moment:
https://github.com/orgs/shorebirdtech/projects/22/views/1

**Comment by KatayR** [comment]
Yesss pleease

**Comment by alexobviously** [comment]
Any update on this? This bug causes me to have to have approximately one confusing conversation every couple of weeks with customer success. It's not like it actually breaks anything but it definitely makes communicating about patch numbers more difficult, especially when we're putting out several patches in a day for critical bugs and I'm trying to explain which version people need to be on for which platform for which bug resolution.

**Comment by eseidel** [comment]
It's on the list.  Pretty high up-voted.  This isn't scheduled for Q4 or Q1, but possible we could get to it in Q2 next year.  Right now we're mostly trying to scale the team to keep up with demand.

https://github.com/shorebirdtech/shorebird/issues/1288 might be another way to solve this, if the notes could contain the desired number.  But yes, it's confusing that Android and iOS have separate patch numbers despite having the same content.

**Comment by eseidel** [comment]
@easymac @nickshorebird I know both of you made some attempts to address this in recent weeks.  Can you update the bug with what you've learned/current status?  Thanks!

**Comment by mml555** [comment]
One question before I offer a patch for this: **does the API support a single patch carrying artifacts for more than one platform?**

If it does, this falls out naturally. Today `shorebird patch --platforms=android,ios` creates a separate patch per platform, each with its own number. If one invocation instead creates one patch and uploads each artifact with its own platform, a multi-platform patch gets one number, one promotion, and one rollback — which is what #1382 is asking for.

While looking at this I hit a related failure mode worth mentioning. Preconditions are asserted at the start of each platform's own turn, so patching `android,ios` from a machine that can't build Apple targets publishes the Android patch and *then* fails, leaving a live patch covering only one platform. Preflighting every platform before building any — and promoting only after all artifacts upload — means a failure anywhere leaves an inert patch instead of a partially-live one. It also removes the double version prompt in #3475.

I have this implemented and passing the existing `shorebird_cli` suite (2086 tests, no new analyzer findings). It restructures `CreatePatchMetadata` so per-platform fields move to a `CreatePatchPlatformMetadata` keyed by platform, which changes the shape of the metadata blob the server receives — so I'd want to know that's acceptable on your side too.

Happy to open a PR if the multi-platform-patch shape is something the API supports. If it isn't, I'll drop it rather than send something that can't work.

**Comment by nickshorebird** [comment]
Hi all - We have done some design and experimentation around unifying the patch numbers in some form and are waiting on bandwidth to implement. No date to share yet.

@mml555 - Thanks for asking before opening the PR. Hold off for now. The shape of the fix is still being finalized on our end, so a multi platform patch is not something I can confirm we would take. Do not want you spending time on something we may not be able to merge. I have also opened #3869 to track the related issue you ran into. Thanks for bringing this up.

**Comment by mml555** [comment]
Understood, holding off. Thanks for answering quickly, and for saying so directly; better to know now than after a PR.

I will keep the change local to our fork and follow this issue for whatever shape you land on. If it is useful later, I am happy to share what we ran into implementing it, but I will wait for your design rather than pre-empt it.

Thanks also for filing #3869. I have added the specific path I hit over there, since it turned out to be a different trigger from the one in your description.

**Comment by eseidel** [comment]
We partially rolled out a fix for this and then rolled back for now (was a bit more complicated than we thought).  Still plan to get back to this as soon as we can, it's high on Mac's list.
